### DISCUSSION TITLE
Missing distinction between control and data transfer calls in the IDS Communication

### DISCUSSION THREAD

**Discussion by sebbader** [International-Data-Spaces-Association/IDS-G#45] [Ideas]
Several issues appeared during the implementation efforts facing the communication patterns. The following things are required:

* only *synchronous* message exchange handled in/from *one* component in *one* process at *one* endpoint
-> How to establish context and how to use the context in different components and across different sessions (in time)?
-> What are the state transitions in each defined sequence? (When are we allowed to e.g. forget the history of a communication?)
-> 
-> What is 'context' for us? How do we define it? Where? Which attributes are needed for the context of which sequences? Context can e.g. a negotiation history, a ContractAgreement, session IDs etc.
-> What are requirements coming from streaming data?

* Communication description is driven by the view from the infrastructure components. Should be the other way around, the component interfaces shall be defined based on generic communication patterns.

**Comment by gboege** [comment]
Discussion after Sebastian Bader left:

- To be covered in message sequence. Stopping data transfer after some policies change or human interaction (access only after 5PM, access paused...) ... there need to be some continuous monitoring capabilities discussed, somehow.

Maybe not highest priority at the moment:
- To be covered somehow: How to handle not only access control, but also usage control enforcement on external systems: "delete after 7 days." This cooperative behaviour between control and data plane have to be covered. Maybe in Certification/DAPS Token? ... Maybe there needs to be some minimum connector attributes like: "data plane capabilities able to do X,Y,Z. 
How could this be covered in policies without having to implement 1.000.000 individual use cases?
Would that mean that connectors have to be split into security profiles and control/data-plane capability grade **to be evaluated in policy**?

- Existing connector technology/implementations/investments shall not be invalidated. Can implementations with or without dataplanes can coexist? How?

**Comment by ssteinbuss** [comment]
# Proposal to move ahead with our work

distinguish between different packages:

## Foundation ##
This is the common foundation for all subsequent activities

### foundational standards (references) ###

please list relevant things here that we use that already exist. 

### Information Model ###

One common model that is used in every other package
(dcat + odrl + IDS specifics (participant, connector, contract,...))

List of information models that we use + list of IDS-specific attributes/classes/packages/. 
use dcat
use odrl
use 

### Identities ##
- messages and data types:
- protocols: state machines for message flows and interaction patterns:
- API binding:

### Trust Frameworks ###
- messages and data types:
- protocols: state machines for message flows and interaction patterns:
- API binding:

### Policies (authorization and Policy Description)
- messages and data types:
- protocols: state machines for message flows and interaction patterns:
- API binding:

## Data Sharing (Conector) ##

### Contract Negotiation ###
part of the control plane

- messages and data types:
- protocols: state machines for message flows and interaction patterns:
- API binding:

### Data Transfer ###
part of the data plane. How data is exchanged with focus on communication and not on how the data plane is built. 

- messages and data types:
- protocols: state machines for message flows and interaction patterns:
- API binding:

## Catalog (Publish and query meta-data) ##

- messages and data types:
- protocols: state machines for message flows and interaction patterns:
- API binding:

## Registration ##

- messages and data types:
- protocols: state machines for message flows and interaction patterns:
- API binding:

## Audit logging ##
currently out of scope 

## Vocabularies ##
currently out of scope

# part of the system layer in RAM # 
What has to be realized into a solution and not part of the communication:
- Policy enforcement would be part of the data plane to be realized and is not part of the communication
- Cyber security assessments are part of the system architecture
- Monitoring of policies.

**Reply by sebbader** [reply]
Catalog: Messages and sequence diagrams (https://github.com/International-Data-Spaces-Association/IDS-G/tree/main/Communication/sequence-diagrams/data-connector-to-metadata-broker) but no state machine yet.
API Bindings are in chapters of the protocol binding markdowns: e.g. https://github.com/International-Data-Spaces-Association/IDS-G/blob/main/Communication/protocols/multipart/README.md#415-Connector-Update and https://github.com/International-Data-Spaces-Association/IDS-G/blob/main/Communication/protocols/ids-rest/README.md#ConnectorUpdateMessage

**Reply by ssteinbuss** [reply]
With regards to the data exchange protocol, we need to define a "meta-protocol" that can be implemented with tangible protocols like HTTP multipart, IDSCP 2, IDS-LDP/REST

**Comment by gbrost** [comment]
Just as a pointer to some work that we did on the IDSCP2. 
We designed a FSM for our protocol that embeds remote attestation and DAPS interaction: https://github.com/industrial-data-space/idscp2-jvm/wiki/IDSCP2-Core
We tried to separate protocol and transport by choosing a modular design:
https://github.com/industrial-data-space/idscp2-jvm/wiki/IDSCP2-Overview
This allows for using arbitrary transports and to integrate other identity schemes.

After establishing the trusted connection, you could either do agreement on the data plane communication or run IDS message exchange directly. 
Even if this probably won't be a solution for everything, maybe we can take some concepts and ideas into our discussion. 
There is also a nice overview of the current IDS protocol variants:
https://docs.google.com/spreadsheets/d/1if-UT9icoAk6czR1LLKvmPGfGH5zI2PpqEbXoKKy8PM/edit#gid=0
